### DISCUSSION TITLE
Use SWC for codemods?

### DISCUSSION THREAD

**Discussion by lgrammel** [swc-project/swc#1643] [Q&A] [answered]
Hi!

I'm thinking about exploring SWC/Rust for codemods and was wondering if that's a usecase that would make sense?

Right now, I'm using Babel + Recast, and in particular I'm curious if something like Recast (retaining most of the formatting during a codemod) is supported / would be easy to add?

Cheers, Lars

**Comment by kdy1** [comment, marked as answer]
I guess it makes sense, although I'm not sure what you mean by `codemod`.
Can you tell me what it is?

**Reply by lgrammel** [reply]
Cool! I meant transformations that changes / upgrade syntax. They are also called codeshifts sometimes. An older example would be transforming JS functions into arrow functions where applicable.

**Reply by Kuranes** [reply]
Would love a codemod that rewrite code like migrates a whole codebase to a specific ecmaversion (say when IE dies, or your website user browser are now all sufficently recent) so that you gain some build time.
As an external tool that can be integrated into IDE like "[import-js"](https://github.com/galooshi/import-js) that could be run on codebase, so that it's done once and for all instead that at each build/pack ?

My most wanted feature would be an transformer helping treeshaking
```javascript
import defaultExport  from "module-name";
defaultExport.export1();
```
=>
```javascript
import { export1 , export2 } from "module-name";
export1();
```

Would allow for Faster (done once and not at each build) and more secure treeshaking, as you can test/check side effect it in dev VS test it on prod build ?

**Reply by kdy1** [reply]
> My most wanted feature would be an transformer helping treeshaking

I think all widely used bundler, including one of swc, will only include `export1` in the bundle. As you can statically know the fact that only `export1` is used by looking at code, the analyzer of the bundler will also same.
And usage analysis is not a slow part. Optimizing it would not affect build time a lot. (Note that maybe some js bundlers can be slow)

I'm sorry but I failed to get what you are trying to say.

**Reply by Kuranes** [reply]
for instance wepback provides with a https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free to note that some cannot be tree shaken.

many "huge old code base maintained by huge number of people" just fail short of being tree shakable because of that so it's all marked with "side effect". 

Doing it as a codemod would allow to do it on a code base as an iterative semi auto process, file by file or folder by folder, until you get to the one that really has side effect and you can mark as such.

Hoping it makes more sense ?

**Reply by kdy1** [reply]
Yes, I think it makes sense.

**Reply by lgrammel** [reply]
I've looked a bit into this, and from what I understand, SWC a primarily a compiler (meaning it transforms human readable source code to minimal machine consumed code) - please correct me if I'm wrong.

Codemods need to transform human readable code to human readable code (ideally with as little formatting changes as possible). This e.g. means that the original token locations need to be retained in the AST for minimal-change reformatting.

I think SWC is awesome and I would love to use in particular the parser piece and if possible also formatting, but I'm not sure if it supports the reformatting / retaining format information piece?

**Reply by kdy1** [reply]
If you are using swc from rust, writing a human-readable code generator is not a hard task.
But if you are using swc from js, ideal codemod is currently impossible because the code generator is writter for a compiler, although retaining original source location may worth (#1131)

**Reply by lgrammel** [reply]
Awesome! The direction I'm thinking of is to write the codemod in Rust. What would be involved in generating human-readable code (ideally with as much of the original formatting as possible)?

**Reply by kdy1** [reply]
All information related to original source code is stored at [swc_common::SourceMap](https://rustdoc.swc.rs/swc_common/struct.SourceMap.html). 
[Span](https://rustdoc.swc.rs/swc_common/struct.Span.html) is a pointer for the data and has a document on it.


For the modification part, [Fold](https://rustdoc.swc.rs/swc_ecma_visit/trait.Fold.html) will be useful. There are lots of transforms using `Fold` at `/ecmascript/transforms` so you can refer to it to see what can be done.

The codegen will require getting original line and column. You can do it with methods on the `SourceMap`. One example is [lookup_char_pos](https://rustdoc.swc.rs/swc_common/struct.SourceMap.html#method.lookup_char_pos).
After getting original line and column, you can print them carefully. (You can even just print original source code if it's not modifed.)

**Reply by lgrammel** [reply]
Thanks for the pointers! I'll see if I can get some trivial codemod working.

**Reply by 0xdevalias** [reply]
@lgrammel Did you ever have any luck getting this going?

**Reply by lgrammel** [reply]
@0xdevalias never got around to explore this

**Reply by Viijay-Kr** [reply]
@lgrammel Did you get to explore this ? I am also in the pursuit of writing a codemods using SWC AST . Would love to see some explorations for reference
